bench-tps/src/log_transaction_service.rs

```diff
@@ -0,0 +1,508 @@
+                    .latest()
+                    .expect("valid timestamp")
+            }),
+            num_bench_tps_transactions,
+            total_num_transactions,
+            bench_tps_cu_consumed,
+            total_cu_consumed,
+        };
+        let _ = block_log_writer.serialize(block_data);
+    }
+
+    fn flush(&mut self) {
+        if let Some(block_log_writer) = &mut self.log_writer {
+            let _ = block_log_writer.flush();
+        }
+    }
+}
+
+#[derive(Clone, Serialize)]
+struct TransactionData {
+    pub blockhash: Option<String>,
+    pub slot_leader: Option<String>,
+    pub signature: String,
+    pub sent_at: Option<DateTime<Utc>>,
+    pub confirmed_slot: Option<Slot>,
+    pub block_time: Option<DateTime<Utc>>,
+    pub successful: bool,
+    pub error: Option<String>,
+    pub timed_out: bool,
+    pub compute_unit_price: u64,
+}
+
+struct TransactionLogWriter {
+    log_writer: Option<CsvFileWriter>,
+}
+
+impl TransactionLogWriter {
+    fn new(transaction_data_file: Option<&str>) -> Self {
+        let transaction_log_writer = transaction_data_file.map(|transaction_data_file| {
+            CsvFileWriter::from_writer(
+                File::create(transaction_data_file).expect("File can be created."),
+            )
+        });
+        Self {
+            log_writer: transaction_log_writer,
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn write(
+        &mut self,
+        blockhash: Option<String>,
+        slot_leader: Option<String>,
+        signature: &Signature,
+        sent_at: DateTime<Utc>,
+        confirmed_slot: Option<Slot>,
+        block_time: Option<i64>,
+        meta: Option<&UiTransactionStatusMeta>,
+        timed_out: bool,
+        compute_unit_price: Option<u64>,
+    ) {
+        let Some(transaction_log_writer) = &mut self.log_writer else {
+            return;
+        };
+        let tx_data = TransactionData {
+            blockhash,
+            slot_leader,
+            signature: signature.to_string(),
+            sent_at: Some(sent_at),
+            confirmed_slot,
+            block_time: block_time.map(|time| {
+                Utc.timestamp_opt(time, 0)
+                    .latest()
+                    .expect("valid timestamp")
+            }),
+            successful: meta.as_ref().map_or(false, |m| m.status.is_ok()),
+            error: meta
+                .as_ref()
+                .and_then(|m| m.err.as_ref().map(|x| x.to_string())),
+            timed_out,
+            compute_unit_price: compute_unit_price.unwrap_or(0),
+        };
+        let _ = transaction_log_writer.serialize(tx_data);
+    }
+
+    fn flush(&mut self) {
+        if let Some(transaction_log_writer) = &mut self.log_writer {
+            let _ = transaction_log_writer.flush();
+        }
+    }
+}
+
+const NUM_RETRY: u64 = 5;
+const RETRY_EVERY_MS: u64 = 4 * DEFAULT_MS_PER_SLOT;
+
+fn call_rpc_with_retry<Func, Data>(f: Func, retry_warning: &str) -> Result<Data>
```

Comment:
This seems a little funky to have buried in log_transaction_service. Is there anywhere else this function might make sense to live?